### PR TITLE
joern-export: SplitByMethod

### PR DESCRIPTION
**Pull request by mpollmeier**
implements https://github.com/joernio/joern/issues/1893
depends on https://github.com/ShiftLeftSecurity/overflowdb/pull/331 and https://github.com/ShiftLeftSecurity/overflowdb/pull/332 - won't compile until upgrade train has completed

Also includes some refactorings - most notably: Representation|Format: changed to uppercase 
(as per Scala convention and support in match clauses), but match on name ignore case when parsing config

### CODE REVIEW

File: joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala [hunk 1]
```diff
@@ -5,35 +5,55 @@ import better.files.File
...
-    format: Format.Value = Format.dot
+    repr: Representation.Value = Representation.Cpg14,
+    format: Format.Value = Format.Dot
   )
 
   /** Choose from either a subset of the graph, or the entire graph (all).
     */
   object Representation extends Enumeration {
-    val ast, cfg, ddg, cdg, pdg, cpg14, all = Value
+    val Ast, Cfg, Ddg, Cdg, Pdg, Cpg14, All, SplitByMethod = Value
```
Comment (fabsx00) [inline joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala:36]: I guess it's a bit strange that ASTs, CFGs, Ddgs, ... are also split by method, and then there's a `SplitByMethod`, but I also can't think of a different way of doing this right now. I guess as long as the docs explain this, we're fine.

**Review by fabsx00 — APPROVED** [review]

File: joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala (hunk 1 above)
Comment (mpollmeier) [reply joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala:36]: discussed via slack: we'll rename `SplitByMethod` to `Cpg`
...  there's no optimal solution 🤷🏻

File: joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala [hunk 2]
```diff
@@ -156,13 +159,46 @@ object JoernExport extends App {
...
+    *   utility map to ensure we don't override output files for identical method names
+    */
+  private def sanitizedFileName(
+    methodName: String,
+    methodFilename: String,
+    fileExtension: String,
+    windowsFilenameDeduplicationHelper: mutable.Set[String]
+  ): String = {
+    val sanitizedMethodName = methodName.replaceAll("[^a-zA-Z0-9-_\\.]", "_")
+    val isWindows           = System.getProperty("os.name").toLowerCase.contains("win")
```
Comment (max-leuthaeuser) [inline joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala:178]: There is `scala.util.Properties.isWin` for this.

Comment (mpollmeier) [reply joern-cli/src/main/scala/io/joern/joerncli/JoernExport.scala:178]: nice, thanks!

**Comment by mpollmeier** [comment]
@max-leuthaeuser I had to lower the 'expected export file count' for windows: https://github.com/joernio/joern/pull/1902/files#diff-8857d40c75e3a90bdfd9c46e94a01e8662591f4d6edc7f3c58bce4077fb58523R19-R25

For some reason the cpg produced on windows is much smaller. I didn't look into the details and for this PR it doesn't really matter - I'm just wondering if that's a known thing.

File: joern-cli/src/test/scala/io/joern/joerncli/JoernExportTests.scala [hunk 3]
```diff
@@ -15,7 +15,14 @@ class JoernExportTests extends AnyWordSpec with Matchers with AbstractJoernCliTe
       val tempDir = os.temp.dir(prefix = "joern-export-test")
       JoernExport.exportCpg(cpg, Representation.Cpg, JoernExport.Format.Graphson, tempDir.toNIO)
       val exportedFiles = os.walk(tempDir).filter(_.toIO.isFile)
-      exportedFiles.size should be >= 100
+
+      // c frontend creates a different cpg on windows - not sure about the details, but purely for the export it doesn't
```
Comment (max-leuthaeuser) [inline joern-cli/src/test/scala/io/joern/joerncli/JoernExportTests.scala:19]: The windows GH action runner does not provide a gcc installation. Thus, scanning system header files results in 0 files. You may fix this by running the c2cpg with `--no-include-auto-discovery`. Maybe we should make this the default as it pops up every now and then (leads to different CPGs on different environments all the time).

Comment (max-leuthaeuser) [reply joern-cli/src/test/scala/io/joern/joerncli/JoernExportTests.scala:19]: @mpollmeier There is https://github.com/joernio/joern/pull/1909 for this.

Comment (mpollmeier) [reply joern-cli/src/test/scala/io/joern/joerncli/JoernExportTests.scala:19]: 👍🏻 I see
re https://github.com/joernio/joern/pull/1909 i'll drop the distinction in this test then

**Pull request merged**